### ISSUE TITLE
Scripting: Allow specifying layer types for which a Tool should be enabled

### ISSUE THREAD

**Issue by eishiya** [mapeditor/tiled#3248]
Scripted Tools can be set to enabled or disabled, but in most cases, this just needs to react to the currently selected layer(s) without much complex logic. Connecting to and disconnecting from layer changed signals as maps change is makes for a lot of tedious boilerplate. It would be great to offload this work to Tiled and just specify what kinds of layers the tool should be enabled or disabled on. Tiled already handles similar logic with Tool.usesSelectedTiles, which prevents the tool from changing when the user picks a tile in a tileset.

In #3246, it was suggested to have boolean properies like `allowForObjectLayers`, `allowForTileLayers`. I think it might be better to specify a single property like `activateOn: [ObjectGroup, ImageLayer, GroupLayer]`, so that its absence can less ambiguously mean "active for all layers".

It would be cool to have an additional property that decides whether to activate the tool when *any* of the selected layers are valid vs requiring *all* selected layers to be valid. Or I guess there can be one default behaviour, and if someone wants it different, they can override it in their script using the usual (tedious) method.

If Tileset tools are ever added, this could be expanded with `Tileset` as an additional acceptable item.

**Comment by bjorn** [comment]
> It would be cool to have an additional property that decides whether to activate the tool when _any_ of the selected layers are valid vs requiring _all_ selected layers to be valid.

Hmm, for most existing tools, they actually base their enabled state only on the "current layer", not the "selected layers". I think it would make sense to do the same for this setting, or if we choose a different behavior for this setting, to do the same for the existing tools (but I'm currently not sure if that's beneficial or which problems that might introduce).

**Comment by eishiya** [comment]
I agree that it should be consistent for scripted and built-in tools. Currently, most of Tiled's tools only work on either one layer (often, but not always, the current layer) or on *any* relevant layer, so making them activate based on selected layers wouldn't make sense. I'd *like* to say I'd love to see some of those tools modified to make sense for working on multiple selected layers, but that just doesn't have very clear logic as to how it should work. For example, it'd be great to Select Same Tile from several layers, but which tile is chosen when you click?

So, in the interest of consistency, until/unless more of Tiled's built-in tools can benefit from being selected based on selected layers rather than current layer, **I'm fine with it being based on the current layer**.

That said, I think the behaviour of setting the current layer could be improved. It's currently very easy to end up in a situation where you have only Tile Layers selected but can only choose Object tools because the current layer is an Object Layer or vice versa, if you deselect layers by modifier+clicking on them. Whenever possible, I think the current layer should be one of the selected layers.

**Comment by bjorn** [comment]
> It's currently very easy to end up in a situation where you have only Tile Layers selected but can only choose Object tools because the current layer is an Object Layer or vice versa, if you deselect layers by modifier+clicking on them. Whenever possible, I think the current layer should be one of the selected layers.

Yeah, actually it tries to do that, so possibly the deselection case is a bit of a bug. The question is how to address it. I guess we could try switching the current layer to the next selected layer, when available.

Thanks for your feedback!
